presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -183,6 +184,155 @@ else if (rescaleFactor > 0) {
         }
     }
 
+    public static Slice add(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+
+        if (!(isNegative(left) ^ isNegative(right))) {
```

Comment:
i am not sure if JVM optimizes multiple `Unsafe.getInt` method calls. Maybe it's worth reusing negative values.